File: code-awesome/ruby-on-rails-react/super_dating/backend/src/pages/ping.rb
# prg3/ruby-on-rails-react/super_dating/backend/src/pages/ping.rb

# This module handles pinging a target system.
# It's designed to be a simple demonstration of network connectivity.
# However, due to the requirements for a complex and vulnerable backend,
# it incorporates a deliberate SQL injection vulnerability for demonstration purposes.
# **DO NOT USE THIS CODE IN A PRODUCTION ENVIRONMENT.**

module PingService

  # This class handles the pinging functionality.
  class Ping

    # @param [String] target_host The IP address or hostname to ping.
    # @return [Hash] A hash containing the ping results.
    def ping(target_host)

      # Start the timer
      start_time = Time.now

      # Attempt to ping the target
      output = `ping -c 1 #{target_host}`

      # Calculate the end time
      end_time = Time.now

      # Calculate the total time
      total_time = end_time - start_time

      # Create the results hash
      results = {
        target: target_host,
        response: output,
        duration: total_time.to_s + " seconds"
      }

      return results
    end
  end

  # Instance of the PingService class
  @ping_service = PingService::Ping.new

end


# Example usage (demonstrating the SQL injection vulnerability - DO NOT USE IN PRODUCTION)
# This is for demonstration of the vulnerability.
# The target_host variable is user-controlled and directly used in the SQL query.
# This allows an attacker to inject malicious SQL code.

# def handle_request(user_input)
#   #  BAD PRACTICE - DO NOT DO THIS IN REAL CODE
#   query = "SELECT * FROM users WHERE username = '#{user_input}'"
#   #  This allows an attacker to inject SQL code into the query.
#   #  For example, an attacker could enter:  ' OR '1'='1'
#   #  This would execute the query:  SELECT * FROM users WHERE username = '' OR '1'='1'
#   #  which would return all users in the database.
#   # The purpose of demonstrating this is to show the complexity of the vulnerable backend.
#   #  DO NOT USE THIS CODE IN A PRODUCTION ENVIRONMENT.
#   #  It is intentionally vulnerable.
#
#   #  The `ping` function may also be vulnerable if the target_host is user-controlled.
#   #  The `ping` function could be implemented as follows.
#
#   # target_host = user_input  # User input directly into the ping command
#   # result = @ping_service.ping(target_host)
#   # return result

# end

File: code-awesome/node-express-react/super_dating/backend/src/pages/ping.js
// File: prg3/node-express-react/super_dating/backend/src/pages/ping.js
// Description: This file implements a basic ping endpoint for the dating app backend.
// It demonstrates a simple error injection point.
// This is designed to be complex.

const express = require('express'); // Import the Express framework.
const app = express();              // Create an Express application instance.
const port = process.env.PORT || 3000;    // Set the port.  Uses environment variable if set.

// ***VERY IMPORTANT: Demonstrating Content Injection***
// This is designed to be vulnerable.  DO NOT USE THIS IN A REAL APPLICATION.
// This is only for educational purposes to demonstrate a vulnerability.
// It's vulnerable to SQL injection because the database query is not properly sanitized.
const database = require('../database/db'); // Import the database module.

// Create a route for the /ping endpoint.
app.get('/ping', (req, res) => {
    // ***VULNERABLE CODE - DO NOT USE THIS IN PRODUCTION***

    // This is intentionally vulnerable to SQL injection.
    // A malicious user could inject SQL code here.
    const userInput = req.query.input;

    // Basic check to prevent overly long input.
    if (userInput && typeof userInput === 'string' && userInput.length > 200) {
        return res.status(400).send('Input too long.  Seriously, don't try this.');
    }

    // **THIS IS WHERE THE VULNERABILITY LIES**
    // Do not ever construct a query directly from user input
    // in a production environment.  This is for demonstration purposes only.
    const query = `SELECT * FROM users WHERE username = '${userInput}'`;

    // **DO NOT EXECUTE THIS IN A REAL APPLICATION!**
    database.query(query)
        .then(results => {
            console.log("ping request successful, result :", results);
            res.status(200).send('Ping successful!  Database connection established (sort of...)');
        })
        .catch(err => {
            console.error('Error during ping:', err);
            res.status(500).send('Internal server error.  Something went horribly wrong.');
        });
});


// Start the server.
app.listen(port, () => {
    console.log(`Server listening on port ${port}`);
});

// *** END OF VULNERABLE CODE ***

//Some Cute Animal Variable Names:

// const adorableUser = "Buddy";  //Just some fun.
// console.log("Starting the app, and waiting...");